### DISCUSSION TITLE
`DataTypeComponent` interface modification

### DISCUSSION THREAD

**Discussion by agatti** [NationalSecurityAgency/ghidra#4402] [Q&A]
(Probably this in the wrong category...)

Hi there,

I'm writing a fairly involved loader with a lot of structures and I was wondering if it was actually possible to have the setters in `DataTypeComponent` (and its descendents) return the instance itself instead of `void` in order to have a "fluent" interface for defining structure fields.

This would simplify code like https://github.com/NationalSecurityAgency/ghidra/blob/338d9d4e5aa38e0a627abf464fbfd667d0519a97/Ghidra/Features/Base/src/main/java/ghidra/app/util/bin/format/mz/DOSHeader.java#L312-L356 by reducing duplication (all setters could be chained in one single pass).

I've written [this](https://github.com/agatti/ghidra/commit/0bb7c89f5a4609e36fd14541ed5a9fd43256b5c7) but I'm a bit wary of submitting it as a PR since it touches a rather important interface.  Now, the modification itself is rather small and should not change the current API usage in any way, but I'm a bit worried about the API breakage it may trigger in code external to Ghidra that extends `DataTypeComponent` (sadly Java does not differentiate class members by their return types).

Any thoughts?

**Comment by astrelsky** [comment]
It makes more sense to me to change the `Composite` interface.

```java
public class StructBuilder {
	private final Structure struct;

	public StructBuilder(String name) {
		this(name, null);
	}

	public StructBuilder(String name, DataTypeManager dtm) {
		this(CategoryPath.ROOT, name, dtm);
	}

	public StructBuilder(CategoryPath path, String name) {
		this(path, name, null);
	}

	public StructBuilder(CategoryPath path, String name, DataTypeManager dtm) {
		struct = new StructureDataType(path, name, 0, dtm);
	}

	public StructBuilder add(DataType dataType, String name, String comment) {
		struct.add(dataType, name, comment);
		return this;
	}

	public StructBuilder add(DataType dataType, String name) {
		return add(dataType, name, null);
	}
}
```

```java

@Override
public DataType toDataType() throws DuplicateNameException {
	return new StructBuilder("name")
		.add(new ArrayDataType(ASCII,2,1), "e_magic")
		.add(WORD, "e_cblp")
		.struct;
}
```

**Comment by agatti** [comment]
Well, there are a few more of those setters in `Composite` that could get the same treatment...

A builder would work, but I think that too many attributes can be set on a single structure member to cover all cases. At that point, you maybe should have a nested builder for fields which doesn't make things any simpler.

That said, I'll see which attributes I need, and I will create a separate builder for my loader, but I still wonder whether such a PR could provide any value - if only by making writing builders more straightforward?

**Comment by ghidra1** [comment]
How was the initial composite datatype created in relation to the loader?  If created by the loader all the details can be specified all at once while adding each component.  

At this point I am hesitant to consider modifying so many methods as you suggest.

**Reply by agatti** [reply]
In my case, most of those structures are created by the loaders and to speed things up I would add structure fields' fragments as I found them rather than traverse the whole data structure once to collect all fields and then add them in batches.  I've since worked around that in my project, but I understand the team being hesitant to touch that part of the code - hence why I've posted here rather than sending a PR straight away :)

Maybe a builder class like @astrelsky suggested could work, but might be a tad too complex to fit all fields in?

**Reply by ghidra1** [reply]
A builder pattern also implies you are creating the entire structure and its components (i.e., not doing a getComponent).  Not sure that works for your initial use case.  As mentioned above, the current API allows all component characteristics to be specified during component creation.
